### PR TITLE
Use custom error pages

### LINKED ISSUE #682: 404 and 500 pages don't load assets
The 404 and 500 pages don't look right in production because the unhashed assets are not available.

The pages should be precompiled as described at 
http://stackoverflow.com/a/15068268/1564027 and
https://www.icelab.com.au/notes/precompiled-rails-static-404-and-500-pages/

I don't know when I will get to this

### PR DESCRIPTION
**Pull request by damianhxy**
### Summary generated by Reviewpad on 24 Jun 23 16:20 UTC
This pull request includes changes to multiple files, which involve removing `rescue_from` blocks for `ActionView::MissingTemplate` exceptions, adding custom error pages using new `error_404` and `error_500` methods, changing authorization levels for some controller actions, modifying before and after action settings, rearranging some skip_before_action statements, and adding new configurations to use custom routes for error pages in the development and production environments. In addition, some unnecessary `get` routes were removed and new routes using `get` were added. Finally, some methods were added and removed from certain files such as `home_controller_spec.rb`, `admins_controller.rb`, and `scoreboards_controller.rb`.

## Description

- Remove the use of `rescue_from ActionView::MissingTemplate` since we have `rescue_from Exception, with: :render_error` under `application_controller.rb`
- Add an `error_500` method to `home_controller` for routing purposes
- Rename `error.html.erb` to `error_500.html.erb`
- Add routes for `/404` and `/500`
- Add `config.exceptions_app = self.routes` to environments, so that `routes.rb` is used for `404` and `500` errors
- Use `render` instead of `redirect_to` in `set_course` so that the url is preserved
- Update tests for error pages

References
- https://stackoverflow.com/a/26286472
- https://medium.com/ruby-on-rails-web-application-development/custom-400-500-error-pages-in-ruby-on-rails-exception-handler-3a04975e4677
- https://web-crunch.com/posts/custom-error-page-ruby-on-rails
- https://dev.to/ayushn21/custom-error-pages-in-rails-4i43

## Motivation and Context

Currently, 404 errors are implicitly handled by Rails rendering the `404.html` page. Not only does this look bad because it doesn't include the layout, but also assets (favicon, image) do not properly load because they use absolute paths.

This PR shifts towards using the existing `error_404` error page that was previously used for `MissingTemplate` errors.

Closes #682

## How Has This Been Tested?

Tested using docker compose install (checkout this PR in the autolab subdirectory and rebuild). It'll help to add a `throw Exception` on some route for testing later, e.g. `assessments_controller#show`.

Note, this PR changes `production.rb.template` by adding the line `config.exceptions_app = self.routes`. However, this doesn't apply to existing instances of Autolab. Thus, the `404` and `500` files are left in for backwards compatibility.

Below is a comparison of the various error pages with and without this additional line. Basically, the only difference is that 404 errors (outside of controllers that use `set_course`) work properly

### Currently on Autolab prod / nightly
**Invalid course name**: e.g. {domain}/courses/dummy/assessments
Screenshot 2023-06-09 at 01 14 47

**Invalid page**: e.g. {domain}/a/b
Screenshot 2023-06-09 at 01 14 55

**Exception**
(Not tested, but should be the same as the following)

### This PR: Before adding `config.exceptions_app = self.routes`
**Invalid course name**: e.g. {domain}/courses/dummy/assessments
Screenshot 2023-06-09 at 00 09 28

**Invalid page**: e.g. {domain}/a/b
Screenshot 2023-06-09 at 00 10 45

**Exception**: on the page you added the Exception to
Screenshot 2023-06-09 at 00 11 00

### This PR: After adding `config.exceptions_app = self.routes`
**Invalid course name**: e.g. {domain}/courses/dummy/assessments
Screenshot 2023-06-09 at 00 20 50

**Invalid page**: e.g. {domain}/a/b
Screenshot 2023-06-09 at 00 20 24

**Exception**: on the page you added the Exception to
Screenshot 2023-06-09 at 00 20 40

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have run rubocop for style check. If you haven't, run `overcommit --install && overcommit --sign` to use pre-commit hook for linting
- [ ] My change requires a change to the documentation, which is located at [Autolab Docs](https://github.com/autolab/docs)
- [ ] I have updated the documentation accordingly, included in this PR

### CODE REVIEW

File: public/404.html [hunk 1]
```diff
@@ -1,3 +1,5 @@
+<!-- This page is obsolete and has been replaced by error_404 -->
```
Comment (najclark) [inline public/404.html:1]: nit: can these files be deleted (`public/404.html`, `public/500.html`)?

**Review by najclark — APPROVED** [review]
lgtm, confirmed behavior works as expected

File: public/404.html (hunk 1 above)
Comment (damianhxy) [reply public/404.html:1]: `500.html` can be safely deleted
`404.html` can be deleted, but unless `config.exceptions_app = self.routes` is added, users will only see a blank page

Comment (damianhxy) [reply public/404.html:1]: In the interest of minimizing dead code, I'll delete the files and bring attention to the change required with the next release. Besides, seeing a broken error page isn't much better than a blank page.

**Pull request merged**